File: front/src/services/order-service.ts
'use client';

import { request } from './api';
import { ApiResponse, PaginationResult } from '@/types/api';
import { OrderVO, OrderCreateDTO, OrderRefundDTO, OrderSearchDTO, InstitutionIncomeVO } from '@/types/order';
import { AxiosResponse } from 'axios';

/**
 * 订单服务 - 处理订单相关API
 */
const orderService = {
  /**
   * 创建订单
   */
  createOrder: async (dto: OrderCreateDTO): Promise<OrderVO> => {
    try {
      const response: AxiosResponse<ApiResponse<OrderVO>> = await request.post('/orders', dto);
      return response.data.data;
    } catch (error) {
      console.error('创建订单失败:', error);
      throw error;
    }
  },

  /**
   * 根据ID获取订单详情
   */
  getOrderById: async (id: number): Promise<OrderVO> => {
    try {
      const response: AxiosResponse<ApiResponse<OrderVO>> = await request.get(`/orders/${id}`);
      return response.data.data;
    } catch (error) {
      console.error(`获取订单详情失败, ID: ${id}:`, error);
      throw error;
    }
  },

  /**
   * 根据订单号查询订单
   */
  getOrderByOrderNo: async (orderNo: string): Promise<OrderVO> => {
    try {
      const response: AxiosResponse<ApiResponse<OrderVO>> = await request.get(`/orders/query?orderNo=${orderNo}`);
      return response.data.data;
    } catch (error) {
      console.error(`查询订单状态失败, 订单号: ${orderNo}:`, error);
      throw error;
    }
  },

  /**
   * 获取订单支付表单
   */
  getPaymentForm: async (orderNo: string): Promise<string> => {
    try {
      const response: AxiosResponse<ApiResponse<string>> = await request.get(`/orders/${orderNo}/payment-form`);
      return response.data.data;
    } catch (error) {
      console.error(`获取支付表单失败, 订单号: ${orderNo}:`, error);
      throw error;
    }
  },

  /**
   * 取消未支付订单
   */
  cancelOrder: async (id: number): Promise<OrderVO> => {
    try {
      const response: AxiosResponse<ApiResponse<OrderVO>> = await request.post(`/orders/${id}/cancel`);
      return response.data.data;
    } catch (error) {
      console.error(`取消订单失败, 订单ID: ${id}:`, error);
      throw error;
    }
  },

  /**
   * 申请退款
   */
  refundOrder: async (id: number, dto: OrderRefundDTO): Promise<OrderVO> => {
    try {
      const response: AxiosResponse<ApiResponse<OrderVO>> = await request.post(`/orders/${id}/refund`, dto);
      return response.data.data;
    } catch (error) {
      console.error(`申请退款失败, 订单ID: ${id}:`, error);
      throw error;
    }
  },

  /**
   * 获取机构收入统计
   */
  getInstitutionIncome: async (): Promise<InstitutionIncomeVO> => {
    try {
      const response: AxiosResponse<ApiResponse<InstitutionIncomeVO>> = await request.get('/orders/institution/income');
      return response.data.data;
    } catch (error) {
      console.error('获取机构收入统计失败:', error);
      throw error;
    }
  },

  /**
   * 处理退款申请（机构管理员或平台管理员）
   */
  processRefund: async (id: number, approved: boolean): Promise<OrderVO> => {
    try {
      const response: AxiosResponse<ApiResponse<OrderVO>> = await request.post(`/orders/admin/${id}/process-refund?approved=${approved}`);
      return response.data.data;
    } catch (error) {
      console.error(`处理退款申请失败, 订单ID: ${id}:`, error);
      throw error;
    }
  },

  /**
   * 高级搜索用户订单
   */
  searchUserOrders: async (searchDTO: OrderSearchDTO): Promise<PaginationResult<OrderVO>> => {
    try {
      const response: AxiosResponse<ApiResponse<PaginationResult<OrderVO>>> = await request.post('/orders/my/search', searchDTO);
      return response.data.data;
    } catch (error) {
      console.error('高级搜索用户订单失败:', error);
      throw error;
    }
  },

  /**
   * 高级搜索机构订单（机构管理员）
   */
  searchInstitutionOrders: async (searchDTO: OrderSearchDTO): Promise<PaginationResult<OrderVO>> => {
    try {
      const response: AxiosResponse<ApiResponse<PaginationResult<OrderVO>>> = await request.post('/orders/institution/search', searchDTO);
      return response.data.data;
    } catch (error) {
      console.error('高级搜索机构订单失败:', error);
      throw error;
    }
  },

  /**
   * 获取机构待处理退款申请（机构管理员）
   */
  getInstitutionPendingRefunds: async (): Promise<OrderVO[]> => {
    try {
      const response: AxiosResponse<ApiResponse<OrderVO[]>> = await request.get('/orders/institution/pending-refunds');
      return response.data.data;
    } catch (error) {
      console.error('获取机构待处理退款申请失败:', error);
      throw error;
    }
  },

  /**
   * 高级搜索所有订单（平台管理员）
   */
  searchAllOrders: async (searchDTO: OrderSearchDTO): Promise<PaginationResult<OrderVO>> => {
    try {
      const response: AxiosResponse<ApiResponse<PaginationResult<OrderVO>>> = await request.post('/orders/admin/search', searchDTO);
      return response.data.data;
    } catch (error) {
      console.error('高级搜索所有订单失败:', error);
      throw error;
    }
  }
};

export default orderService; 

File: src/main/java/com/zhangziqi/online_course_mine/service/OrderService.java
package com.zhangziqi.online_course_mine.service;

import com.zhangziqi.online_course_mine.model.dto.order.OrderCreateDTO;
import com.zhangziqi.online_course_mine.model.dto.order.OrderRefundDTO;
import com.zhangziqi.online_course_mine.model.dto.order.OrderSearchDTO;
import com.zhangziqi.online_course_mine.model.vo.OrderVO;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;
import java.util.Map;

/**
 * 订单服务接口
 */
public interface OrderService {

    /**
     * 创建订单
     *
     * @param courseId 课程ID
     * @param userId 用户ID
     * @return 创建的订单VO
     */
    OrderVO createOrder(Long courseId, Long userId);

    /**
     * 处理支付宝同步回调
     *
     * @param params 支付宝回调参数
     * @return 处理结果
     */
    String handleAlipayReturn(Map<String, String> params);

    /**
     * 处理支付宝异步通知
     *
     * @param params 支付宝通知参数
     * @return 处理结果
     */
    String handleAlipayNotify(Map<String, String> params);

    /**
     * 根据订单号查询订单
     *
     * @param orderNo 订单号
     * @return 订单VO
     */
    OrderVO getOrderByOrderNo(String orderNo);

    /**
     * 根据ID查询订单
     *
     * @param id 订单ID
     * @return 订单VO
     */
    OrderVO getOrderById(Long id);
    
    /**
     * 取消订单
     *
     * @param id 订单ID
     * @param userId 用户ID
     * @return 取消后的订单VO
     */
    OrderVO cancelOrder(Long id, Long userId);
    
    /**
     * 申请退款
     *
     * @param id 订单ID
     * @param dto 退款信息
     * @param userId 用户ID
     * @return 更新后的订单VO
     */
    OrderVO refundOrder(Long id, OrderRefundDTO dto, Long userId);
    
    /**
     * 处理退款（机构管理员或平台管理员）
     *
     * @param id 订单ID
     * @param approved 是否批准
     * @param operatorId 操作人ID
     * @return 更新后的订单VO
     */
    OrderVO processRefund(Long id, boolean approved, Long operatorId);
    
    /**
     * 统计机构总收入（不含退款）
     *
     * @param institutionId 机构ID
     * @return 总收入金额
     */
    java.math.BigDecimal calculateInstitutionTotalIncome(Long institutionId);
    
    /**
     * 统计机构总退款
     *
     * @param institutionId 机构ID
     * @return 总退款金额
     */
    java.math.BigDecimal calculateInstitutionTotalRefund(Long institutionId);
    
    /**
     * 获取机构净收入
     *
     * @param institutionId 机构ID
     * @return 净收入金额（总收入-总退款）
     */
    java.math.BigDecimal calculateInstitutionNetIncome(Long institutionId);
    
    /**
     * 执行支付宝退款
     *
     * @param orderNo 订单号
     * @param refundAmount 退款金额
     * @param refundReason 退款原因
     * @return 退款结果，true表示成功，false表示失败
     */
    boolean executeAlipayRefund(String orderNo, java.math.BigDecimal refundAmount, String refundReason);
    
    /**
     * 处理支付成功回调
     *
     * @param orderNo 订单号
     */
    void handlePaymentSuccess(String orderNo);

    /**
     * 根据搜索条件查询用户订单
     *
     * @param searchDTO 搜索条件
     * @param userId 用户ID
     * @return 分页订单VO
     */
    Page<OrderVO> searchUserOrders(OrderSearchDTO searchDTO, Long userId);
    
    /**
     * 根据搜索条件查询机构订单
     *
     * @param searchDTO 搜索条件
     * @param institutionId 机构ID
     * @return 分页订单VO
     */
    Page<OrderVO> searchInstitutionOrders(OrderSearchDTO searchDTO, Long institutionId);
    
    /**
     * 根据搜索条件查询所有订单（管理员）
     *
     * @param searchDTO 搜索条件
     * @return 分页订单VO
     */
    Page<OrderVO> searchAllOrders(OrderSearchDTO searchDTO);
    
    /**
     * 获取机构待处理退款申请
     *
     * @param institutionId 机构ID
     * @return 待处理退款申请列表
     */
    List<OrderVO> getInstitutionPendingRefunds(Long institutionId);
    
    /**
     * 生成支付表单
     *
     * @param orderNo 订单号
     * @return 支付表单HTML字符串
     */
    String generatePaymentForm(String orderNo);
}